### ISSUE TITLE
Generalise DiceLoss never decrease

### ISSUE THREAD

**Issue by hei6775** [LIVIAETS/boundary-loss#41]
Hello, my project use your losses. I use the generalise dice loss in my segment. I want to explain that my data set has been normalized. The network has a batchnorm layer, and the last layer of the network is conv. The input of GDL is the output of softmax function. lr is 5e-5 Neither dicelos nor gdiceloss will decline. If I replace the loss function with the cross entropy function, it goes down. I looked for some possible reasons, but it didn't work. You have rich experience. Do you have any suggestions? Thank you for your work. Thank you for your work for open source

**Comment by hei6775** [comment]
Hello, I have new something to share with you. In my experiments, I found that the GDL is hard to decrease. According to the Paper "AnatomyNet: Deep Learning for Fast...", the generalized dice loss uses squared volume weights. However, it makes the optimization unstable in the extremely unbalanced segmentation. I've replaced GDL with DICELoss and only focused on the classes of interest, which will decrease as expected. I said before that diceloss didn't work, maybe it was my little mistake. Is it still possible to use boundaryoss and diceloss together? Maybe I'll add FocalLoss, because diceloss is unstable.

**Comment by HKervadec** [comment]
Hello,

Yes, you can use Boundary loss with any other loss; they are all compatible. It is about finding a combination that works for your specific problem. 

My experience with the GDL is rather “_chaotic_”, and I usually prefer other losses (even the simple cross-entropy loss). The focal loss is indeed quite powerful, from my more recent experiments, so I encourage you to give it a go. 

> I've replaced GDL with DICELoss and only focused on the classes of interest, which will decrease as expected. 

How many classes does your problem have? Indeed, supervising different classes differently (either with a different weight, or another loss) is not out of the ordinary. You can also decide to supervise the background explicitely (`idc=[0,1,2...]`, or implicitely (`idc=[1,2,...]`). 

Let me know,

Hoel

**Comment by HKervadec** [comment]
> Hello, I have new something to share with you. In my experiments, I found that the GDL is hard to decrease. According to the Paper "AnatomyNet: Deep Learning for Fast...", the generalized dice loss uses squared volume weights. However, it makes the optimization unstable in the extremely unbalanced segmentation. 

Related to that general topic, there is that pre-print (still under revision) from colleagues that might interest you: 
[The hidden label-marginal biases of segmentation losses](https://arxiv.org/abs/2104.08717)
